File: Byte_Compressor/src/core/byte_compressor.c
/*
 * ByteCompressor - Main API Implementation
 */

#include "byte_compressor.h"
#include "ans_codec.h"
#include "../utils/crc32.h"
#include <stdio.h>
#include <math.h>

const char *bc_version(void) {
    return "ByteCompressor v1.0.0 — Deep Space Communication Codec";
}

/* ── Auto-detect best compression profile ── */

BCProfile bc_auto_detect_profile(const uint8_t *data, size_t size) {
    if (size < 16) return BC_PROFILE_RAW_ANS;

    /* Heuristic 1: Check if data is sequential/slowly changing (telemetry) */
    size_t delta_zeros = 0;
    size_t small_deltas = 0;
    for (size_t i = 1; i < size && i < 4096; i++) {
        int diff = (int)data[i] - (int)data[i - 1];
        if (diff == 0) delta_zeros++;
        if (abs(diff) <= 4) small_deltas++;
    }

    size_t sample = (size < 4096) ? size - 1 : 4095;
    double delta_zero_ratio = (double)delta_zeros / sample;
    double small_delta_ratio = (double)small_deltas / sample;

    /* High sequential correlation → telemetry */
    if (small_delta_ratio > 0.7) return BC_PROFILE_TELEMETRY;

    /* Moderate correlation → image-like */
    if (small_delta_ratio > 0.4) return BC_PROFILE_IMAGE;

    /* Check entropy — very high entropy means random, use raw ANS */
    double entropy = ans_compute_entropy(data, size < 4096 ? size : 4096);
    if (entropy > 7.5) return BC_PROFILE_RAW_ANS;

    /* Default: generic BWT pipeline */
    return BC_PROFILE_GENERIC;
}

/* ── Compress with analysis ── */

BCResult bc_compress_analyze(const uint8_t *input, size_t input_size,
                             uint8_t *output, size_t output_capacity,
                             const BCConfig *config) {
    BCResult result = {0};
    result.original_size = input_size;

    BCConfig cfg;
    if (config) {
        cfg = *config;
    } else {
        cfg = bc_default_config();
        cfg.profile = bc_auto_detect_profile(input, input_size);
    }
    result.profile_used = cfg.profile;

    /* Compute entropy statistics */
    result.entropy = ans_compute_entropy(input, input_size);
    result.theoretical_min = ans_theoretical_min_size(input, input_size);

    /* Compress */
    result.compressed_size = bc_compress(input, input_size, output, output_capacity, &cfg);

    if (result.compressed_size > 0 && input_size > 0) {
        result.compression_ratio = (double)input_size / (double)result.compressed_size;
        result.space_saving = (1.0 - (double)result.compressed_size / (double)input_size) * 100.0;
        result.bits_per_symbol = ((double)result.compressed_size * 8.0) / (double)input_size;

        if (result.entropy > 0.001) {
            result.shannon_efficiency = (result.entropy / result.bits_per_symbol) * 100.0;
            if (result.shannon_efficiency > 100.0) result.shannon_efficiency = 100.0;
        }
    }

    result.block_count = (uint32_t)((input_size + cfg.block_size - 1) / cfg.block_size);

    return result;
}

/* ── Decompress with verification ── */

BCResult bc_decompress_verify(const uint8_t *input, size_t input_size,
                              uint8_t *output, size_t output_capacity) {
    BCResult result = {0};

    result.compressed_size = input_size;
    result.original_size = bc_decompress(input, input_size, output, output_capacity);

    if (result.original_size > 0) {
        result.compression_ratio = (double)result.original_size / (double)input_size;
        result.space_saving = (1.0 - (double)input_size / (double)result.original_size) * 100.0;
        result.entropy = ans_compute_entropy(output, result.original_size);
        result.bits_per_symbol = ((double)input_size * 8.0) / (double)result.original_size;

        if (result.entropy > 0.001) {
            result.shannon_efficiency = (result.entropy / result.bits_per_symbol) * 100.0;
            if (result.shannon_efficiency > 100.0) result.shannon_efficiency = 100.0;
        }
    }

    return result;
}

/* ── Print Report ── */

void bc_print_result(const BCResult *result, const char *filename) {
    printf("\n");
    printf("╔══════════════════════════════════════════════════════════════╗\n");
    printf("║           ByteCompressor — Compression Report              ║\n");
    printf("╠══════════════════════════════════════════════════════════════╣\n");

    if (filename) {
        printf("║  File:              %-40s║\n", filename);
    }

    printf("║  Profile:           %-40s║\n", bc_profile_name(result->profile_used));
    printf("╠══════════════════════════════════════════════════════════════╣\n");

    printf("║  Original size:     %10zu bytes %29s║\n", result->original_size, "");
    printf("║  Compressed size:   %10zu bytes %29s║\n", result->compressed_size, "");
    printf("║  Compression ratio: %10.2fx %31s║\n", result->compression_ratio, "");
    printf("║  Space saving:      %10.1f%% %30s║\n", result->space_saving, "");

    printf("╠══════════════════════════════════════════════════════════════╣\n");
    printf("║  Shannon Entropy & Theoretical Analysis                    ║\n");
    printf("╠══════════════════════════════════════════════════════════════╣\n");

    printf("║  Shannon entropy:   %10.4f bits/symbol %19s║\n", result->entropy, "");
    printf("║  Theoretical min:   %10.1f bytes %25s║\n", result->theoretical_min, "");
    printf("║  Actual bits/sym:   %10.4f bits/symbol %19s║\n", result->bits_per_symbol, "");
    printf("║  Shannon efficiency:%10.1f%% %30s║\n", result->shannon_efficiency, "");
    printf("║  Block count:       %10u %31s║\n", result->block_count, "");

    printf("╠══════════════════════════════════════════════════════════════╣\n");

    if (result->shannon_efficiency > 95.0) {
        printf("║  [EXCELLENT] Near-optimal: within 5%% of Shannon limit     ║\n");
    } else if (result->shannon_efficiency > 85.0) {
        printf("║  [GOOD] Efficient compression, close to theoretical limit  ║\n");
    } else {
        printf("║  [NOTE] Data may have high entropy or complex structure    ║\n");
    }

    printf("╚══════════════════════════════════════════════════════════════╝\n\n");
}

File: Byte_Compressor/src/core/byte_compressor.js
/*
 * ByteCompressor - Main API
 * Deep Space Communication Data Compression System
 */

const { computeEntropy, theoreticalMinSize } = require('./ans_codec');
const { PROFILE, profileName, compressFile, decompressFile, DEFAULT_BLOCK_SIZE } = require('./block_codec');
const { crc32 } = require('./crc32');

const VERSION_STRING = 'ByteCompressor v3.0.0 — Deep Space Communication Codec';

function autoDetectProfile(data) {
    if (data.length < 16) return PROFILE.RAW_ANS;

    const sampleSize = Math.min(data.length, 8192);

    // 1. Check if data is text/code (printable ASCII + UTF-8)
    let printable = 0;
    for (let i = 0; i < sampleSize; i++) {
        const b = data[i];
        if ((b >= 0x20 && b <= 0x7E) || b === 0x09 || b === 0x0A || b === 0x0D) {
            printable++;
        }
    }
    const textRatio = printable / sampleSize;

    // 2. Check sequential/telemetry patterns
    let smallDeltas = 0;
    for (let i = 1; i < sampleSize; i++) {
        if (Math.abs(data[i] - data[i - 1]) <= 4) smallDeltas++;
    }
    const smallDeltaRatio = smallDeltas / (sampleSize - 1);

    // 3. Entropy
    const entropy = computeEntropy(data.subarray(0, sampleSize));

    // Decision tree (optimized for real-world files):

    // High entropy (>7.5): already compressed (JPG, MP4, ZIP, etc.) — minimal overhead
    if (entropy > 7.5) return PROFILE.RAW_ANS;

    // Telemetry: slowly changing values (sensor data, sequential)
    if (smallDeltaRatio > 0.7) return PROFILE.TELEMETRY;

    // Text/code: >80% printable ASCII → LZ77+ANS (best for real-world text/code)
    // LZ77+ANS outperforms LZ77+BWT+MTF+ANS because BWT on LZ77 output
    // disrupts the token structure rather than helping
    if (textRatio > 0.80) return PROFILE.LZ77_ANS;

    // Image-like: moderate sequential correlation
    if (smallDeltaRatio > 0.4) return PROFILE.IMAGE;

    // General data > 2KB: LZ77+ANS for dictionary-based compression
    if (data.length > 2048) return PROFILE.LZ77_ANS;

    // Small structured data: pure BWT
    return PROFILE.GENERIC;
}

function compressAnalyze(input, config = {}) {
    const profile = config.profile !== undefined ? config.profile : autoDetectProfile(input);
    const blockSize = config.blockSize || DEFAULT_BLOCK_SIZE;

    const entropy = computeEntropy(input);
    const theoMin = theoreticalMinSize(input);

    const compressed = compressFile(input, { profile, blockSize });

    const result = {
        originalSize: input.length,
        compressedSize: compressed ? compressed.length : 0,
        compressionRatio: 0,
        spaceSaving: 0,
        entropy,
        theoreticalMin: theoMin,
        shannonEfficiency: 0,
        bitsPerSymbol: 0,
        blockCount: Math.ceil(input.length / blockSize),
        profileUsed: profile,
        compressed
    };

    if (compressed && compressed.length > 0 && input.length > 0) {
        result.compressionRatio = input.length / compressed.length;
        result.spaceSaving = (1 - compressed.length / input.length) * 100;
        result.bitsPerSymbol = (compressed.length * 8) / input.length;
        if (entropy > 0.001) {
            result.shannonEfficiency = Math.min(100, (entropy / result.bitsPerSymbol) * 100);
        }
    }

    return result;
}

function decompressVerify(input) {
    const decompressed = decompressFile(input);

    const result = {
        originalSize: decompressed ? decompressed.length : 0,
        compressedSize: input.length,
        compressionRatio: 0,
        spaceSaving: 0,
        entropy: 0,
        bitsPerSymbol: 0,
        shannonEfficiency: 0,
        decompressed
    };

    if (decompressed && decompressed.length > 0) {
        result.compressionRatio = decompressed.length / input.length;
        result.spaceSaving = (1 - input.length / decompressed.length) * 100;
        result.entropy = computeEntropy(decompressed);
        result.bitsPerSymbol = (input.length * 8) / decompressed.length;
        if (result.entropy > 0.001) {
            result.shannonEfficiency = Math.min(100, (result.entropy / result.bitsPerSymbol) * 100);
        }
    }

    return result;
}

function printResult(result, filename) {
    const lines = [];
    const l = (s) => lines.push(s);

    l('');
    l('\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557');
    l('\u2551           ByteCompressor \u2014 Compression Report              \u2551');
    l('\u2560\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2563');

    if (filename) l(`\u2551  File:              ${filename.padEnd(40)}\u2551`);
    l(`\u2551  Profile:           ${profileName(result.profileUsed).padEnd(40)}\u2551`);
    l('\u2560\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2563');

    l(`\u2551  Original size:     ${String(result.originalSize).padStart(10)} bytes${' '.repeat(25)}\u2551`);
    l(`\u2551  Compressed size:   ${String(result.compressedSize).padStart(10)} bytes${' '.repeat(25)}\u2551`);
    l(`\u2551  Compression ratio: ${result.compressionRatio.toFixed(2).padStart(10)}x${' '.repeat(27)}\u2551`);
    l(`\u2551  Space saving:      ${result.spaceSaving.toFixed(1).padStart(10)}%${' '.repeat(27)}\u2551`);

    l('\u2560\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2563');
    l('\u2551  Shannon Entropy & Theoretical Analysis                    \u2551');
    l('\u2560\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2563');

    l(`\u2551  Shannon entropy:   ${result.entropy.toFixed(4).padStart(10)} bits/symbol${' '.repeat(15)}\u2551`);
    l(`\u2551  Theoretical min:   ${result.theoreticalMin.toFixed(1).padStart(10)} bytes${' '.repeat(21)}\u2551`);
    l(`\u2551  Actual bits/sym:   ${result.bitsPerSymbol.toFixed(4).padStart(10)} bits/symbol${' '.repeat(15)}\u2551`);
    l(`\u2551  Shannon efficiency:${result.shannonEfficiency.toFixed(1).padStart(10)}%${' '.repeat(27)}\u2551`);
    l(`\u2551  Block count:       ${String(result.blockCount).padStart(10)}${' '.repeat(28)}\u2551`);

    l('\u2560\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2563');

    if (result.shannonEfficiency > 95) {
        l('\u2551  [EXCELLENT] Near-optimal: within 5% of Shannon limit      \u2551');
    } else if (result.shannonEfficiency > 85) {
        l('\u2551  [GOOD] Efficient compression, close to theoretical limit  \u2551');
    } else {
        l('\u2551  [NOTE] Data may have high entropy or complex structure    \u2551');
    }

    l('\u255a\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255d');
    l('');

    return lines.join('\n');
}

module.exports = {
    VERSION_STRING,
    PROFILE,
    autoDetectProfile,
    compressAnalyze,
    decompressVerify,
    printResult,
    profileName
};
